### PR TITLE
refactor(rome_analyze): refactor how Visitors, Queryables and QueryMatches are related

### PR DESCRIPTION
**Pull request by leops**
## Summary

This PR modifies how various types related to the `Queryable` trait are injected into the analyzer infrastructure. This includes:
- Allowing `Queryable` types to dynamically register the `Visitor` instances that emits them to the analyzer
- Replacing the `QueryMatch` enum with `TypeId` and `Box<dyn Any>` to allow any arbitrary type to be emitted as a visitor event

The main goal of this refactor is to allow individual rules to declare custom syntax tree visitors to emit the event they rely upon as an optimization opportunity instead of having to traverse the syntax tree in their `run` method. For instance the `useYield` rule could be implemented like this:

<details>
<summary>Code</summary>

```rust
impl Rule for UseYield {
    type Query = MissingYield;
}

struct MissingYield(AnyFunctionLike);

impl Queryable for MissingYield {
    type Input = Self;
    type Output = AnyFunctionLike;
    
    fn build_visitor(
        analyzer: &mut impl AddVisitor<Self::Language>,
        _: &<Self::Language as Language>::Root,
    ) {
        analyzer.add_visitor(Phases::Syntax, MissingYieldVisitor::default());
    }

    fn unwrap_match(services: &ServiceBag, query: &Self::Input) -> Self::Output {
        query.0.clone()
    }
}

#[derive(Default)]
struct MissingYieldVisitor {
    stack: Vec<(AnyFunctionLike, bool)>,
}

impl Visitor for MissingYieldVisitor {
    fn visit(
        &mut self,
        event: &WalkEvent<SyntaxNode<Self::Language>>,
        ctx: VisitorContext<Self::Language>,
    ) {
        match event {
            WalkEvent::Enter(node) => {
                if let Some(node) = AnyFunctionLike::cast_ref(node) {
                    self.stack.push((node, false));
                }
                
                if let Some((_, has_yield)) = self.stack.last_mut() {
                    if JsYieldExpression::can_cast(node.kind()) {
                        *has_yield = true;
                    }
                }
            }
            WalkEvent::Leave(node) => {
                if let Some(exit_node) = AnyFunctionLike::cast_ref(node) {
                    if let Some((enter_node, has_yield)) = self.stack.pop() {
                        assert_eq!(enter_node, exit_node);
                        if !has_yield {
                            ctx.match_query(MissingYield(enter_node));
                        }
                    }
                }
            }
        }
    }
}
```

</details>

## Test Plan

This is an infrastructure-only change, the code should continue to build and existing tests should continue to pass correctly.

## Documentation

- [ ] The PR requires documentation
- [ ] I will create a new PR to update the documentation

### CODE REVIEW

**Comment by xunilrj** [comment]
In the yield example, the visitor will be called for all nodes?
Why not use something like the dispatch we have for rules that only answer ```Ast<...>``` queries that runs on the first pass?
We can give this new trait `&mut` to the service bag.

```rust
struct MissingYields(Vec<JsFunctionDeclaration>);

impl RuleService for MissingYields {
    type Query = Ast<JsFunctionDeclaration | JsYieldExpression>; // just an example

    fn visit(...) {
         match event {
            WalkEvent::Enter(node) => {
              ...
            }
            WalkEvent::Leave(node) => {
                ...
            }
        }
    }

    fn finish(bag: &mut ServiceBag) {
         bag.insert(MissingYields(...));
    .}
}
```

The main difference is that we introduce a hook to services. We can have later: FindReactCallsService, FunctionsWithJSX etc...
We can expand the 

```rust
type Query = SemanticServices
```

To actually support requesting any services:

```rust
type Query = Services<SemanticModel, FindReactCallsService>
```

What do you think?

**Comment by leops** [comment]
For now this is mostly implemented as a simple extension to the existing Visitor API that allows individual rules to run custom specialized visitors on the syntax tree in parallel to existing visitors like the semantic model builder. Being able to specify query / filters on visitors seems interesting but would probably require adding another layer of indirection between the syntax tree preorder iterator and the visitors.

**Comment by MichaReiser** [comment]
!bench_analyzer

**Comment by leops** [comment]
!bench_analyzer

**Comment by ematipico** [comment]
@leops it seems that the action didn't finish

**Comment by leops** [comment]
!bench_analyzer

File: crates/rome_analyze/src/registry.rs [hunk 1]
```diff
@@ -170,53 +199,107 @@ impl<L: Language + Default> RegistryVisitor<L> for RuleRegistryBuilder<'_, L> {
...
         let phase = &mut self.phase_rules[params.phase as usize];
 
-        let rules = match &params.query {
-            QueryMatch::Syntax(node) => {
+        let query_type = params.query.type_id();
+        let Some(rules) = phase.type_rules.get(&query_type) else { return };
+
+        let rules = match rules {
+            TypeRules::SyntaxRules { rules } => {
+                let node = params.query.downcast_ref::<SyntaxNode<L>>().unwrap();
```
Comment (MichaReiser) [inline crates/rome_analyze/src/registry.rs:289]: Do you see a way that would avoid the need for the runtime assertion that `rules` and the `query` type (downcast) match?

**Review by MichaReiser — APPROVED** [review]

File: crates/rome_analyze/src/query.rs [hunk 2]
```diff
@@ -1,91 +1,49 @@
...
-    registry::{NodeLanguage, Phase},
-    services::FromServices,
-    ServiceBag,
-};
+use rome_rowan::{Language, SyntaxKindSet, TextRange};
+
+use crate::{registry::Phase, services::FromServices, Phases, ServiceBag, Visitor};
 
 /// Trait implemented for all types, for example lint rules can query them to emit diagnostics or code actions.
 pub trait Queryable: Sized {
```
Comment (ematipico) [inline crates/rome_analyze/src/query.rs:8]: I suppose this trait should have more documentation now

File: crates/rome_analyze/src/matcher.rs [hunk 3]
```diff
@@ -19,12 +24,52 @@ pub struct MatchQueryParams<'phase, 'query, L: Language> {
...
     pub services: &'phase ServiceBag,
     pub signal_queue: &'query mut BinaryHeap<SignalEntry<'phase, L>>,
     pub apply_suppression_comment: SuppressionCommentEmitter<L>,
 }
 
+/// Wrapper type for a [QueryMatch]
+///
+/// This type is functionally equivalent to `Box<dyn QueryMatch + Any>`, it
+/// emulates dynamic dispatch for both traits and allows downcasting to a
+/// reference or owned type.
```
[the remark this answers is not in the document]
Comment (leops) [reply crates/rome_analyze/src/matcher.rs:37]: Currently the Rust language restricts the use of the `dyn Trait` syntax to only a single trait, with an exception for "marker traits" that do not have any methods like `Send` or `Sync`. So it's possible to write `dyn Any + Send + Sync`, but `dyn Any + QueryMatch` is not allowed as both of these traits declare one or more methods

**Review by ematipico — APPROVED** [review]

**Pull request merged**